File: DataStartFrom20241124/T5.cpp
#include <iostream>
#include <cstdio>
#include <vector>
#include <cstring>
#include <algorithm>
#include <vector>
using namespace std;
#define Test

class Solution {
public:
    string longestPalindrome(string s) {
        int sz = s.size();
        vector<vector<int>> f(sz, vector<int>(sz, 0));
        for(int i = 0;i < sz;i ++)
            f[i][i] = 1;
        for(int i = 0;i < sz;i ++){
            for(int j = i - 1;j >= 0;j --){
                if(j + 1 <= i - 1){
                    f[j][i] = f[j + 1][i - 1] && (s[j] == s[i]);
                }
                else{
                    f[j][i] = (s[i] == s[j]);
                }
            }
        }
        int ans = 1;
        string t = string(1, s[0]);
        for(int i = 0;i < sz;i ++){
            for(int j = i + 1;j < sz;j ++){
                if(f[i][j] && j - i + 1 > ans){
                    ans = j - i + 1;
                    t = s.substr(i, j - i + 1);
                }
            }
        }
        return t;
    }
};

int main(){
    #ifdef Test
        freopen("file.in", "r", stdin);
        freopen("file.out", "w", stdout);
    #endif
    string s = "cbbd";
    Solution example;
    cout << example.longestPalindrome(s);
    return 0;   
}

File: CodeDataStartFrom20250808/src/problems_set/T5.java
package problems_set;
public class T5 {
    public String longestPalindrome(String s) {
        // 甚至我感到很奇怪，这道题为什么不是贪心，而是动态规划
        // 时间复杂度估计是O(n^2)
        // 好像想起来了，列的是展开的动态方程

        // 1. dp解法
        // 枚举需要耗费O(n^2)的时间
        // 由此的，判定字符串是否为回文的时间限制为O(1)
        char[] sc = s.toCharArray();
        int ans = 1, lf = 0, ri = 0, n = sc.length;
        for (int i = 0;i < n;i ++) {
            int l = i, r = i;
            while(l - 1 >= 0 && r + 1 < n && sc[l - 1] == sc[r + 1]) {
                -- l;
                ++ r;
            }
            if(r - l + 1 > ans) {
                ans = r - l + 1;
                lf = l;
                ri = r;
            }
        }
        for (int i = 0;i < n - 1;i ++) {
            int l = i, r = i + 1;
            if(sc[l] == sc[r]) {
                while(l - 1 >= 0 && r + 1 < n && sc[l - 1] == sc[r + 1]) {
                    -- l;
                    ++ r;
                }
                if(r - l + 1 > ans) {
                    ans = r - l + 1;
                    lf = l;
                    ri = r;
                }
            }
        }
        return s.substring(lf, ri + 1);
    }
}